### PR TITLE
SITES-947: Re-factor parsing of sites.json to get the numerical site_id value

### PR DESCRIPTION
**Pull request by jbickar**
# READY FOR REVIEW

# Summary
- Dramatically reduce the time and resources that it takes to parse `sites.json` to get the `site_id` value

# Needed By (Date)
- ASAP

# Criticality
- How critical is this PR on a 1-10 scale?  10/10
- Significantly affects migration time

# Steps to Test

1. Use [this inventory](https://github.com/SU-SWS/ansible-playbooks/files/2439710/PR62.txt)
2. Migrate to `dev`, `test`, and `prod` with `full-migration-playbook.yml`
3. Ensure that `setup-site` step is skipped if the site exists on the environment
4. Ensure that caches are cleared and vhosts/custom domains are added
5. Delete one or both of the sites on one environment
6. Ensure that the site is created successfully
7. Migrate to one environment using `migration-sync-only-playbook.yml`
8. Run `post-db-restore-playbook.yml`
9. Run `add-vhost-playbook.yml`
10. Try to think of any edge cases that this approach would **not** work on

# Affected Projects or Products
- ACSF

# Associated Issues and/or People
- JIRA ticket: https://stanfordits.atlassian.net/browse/SITES-947
- Any other contextual information that might be helpful:
    - I believe that this approach should be sufficient to get us a unique `site_id`. See comments in the `sites-json` role.
    - There is probably a more elegant way to do this leveraging Ansible (rather than shell commands like `grep` and `sed`, but this is a significant performance improvement, particularly when `sites.json` gets to be very large when there are many sites on a Factory.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: roles/add-vhost/tasks/main.yml [hunk 1]
```diff
@@ -57,19 +57,3 @@
     register: inventory_hostname_domain_response
   when:
     inventory_hostname != acsf_site_name
-
-# Ex. sunetid.people.stanford.edu
-- name: Add SUNetID as custom domain
```
Comment (jbickar) [inline roles/add-vhost/tasks/main.yml]: This removal is tangential to the substance of this PR. I just noticed that we don't need it anymore now that we have the Customer Standard Domain of people.stanford.edu for the leland stack.

Comment (sherakama) [reply roles/add-vhost/tasks/main.yml]: Fine by me.

File: roles/sites-json/tasks/main.yml [hunk 2]
```diff
@@ -8,145 +8,70 @@
...
-#
-# The array will be output in alphabetical order (id, site, url), but as the keys
-# are object properties, the order does not matter.
-#
-- name: debug sites_formatted
-  debug:
-    var: sites_formatted
-  when: print_debug_messages == "TRUE"
+# sites_json_raw will be a really long string like this:
+# "jbickar.cardinalsites.acsitefactory.com": {"name":"sbiybj3061","flags":{"preferred_domain":true},"conf":{"gardens_site_id":3061,"gardens_db_name":"sbiybj3061","acsf_site_id":3061,"acsf_db_name":"sbiybj3061","site_api_key":"fc28cad31e56b5ae1e97629348872878276510bc"}},
```
Comment (sherakama) [inline roles/sites-json/tasks/main.yml]: Probably don't want to publish these ids and keys.

**Comment by sherakama** [comment]
```
failed: [academicskills] (item=sed -i 's/^.*acsf_site_id/acsf_site_id/g' /tmp/academicskills/sites_json_raw) => {"changed": true, "cmd": "sed -i 's/^.*acsf_site_id/acsf_site_id/g' /tmp/academicskills/sites_json_raw", "delta": "0:00:00.008569", "end": "2018-10-02 17:13:12.127042", "item": "sed -i 's/^.*acsf_site_id/acsf_site_id/g' /tmp/academicskills/sites_json_raw", "msg": "non-zero return code", "rc": 1, "start": "2018-10-02 17:13:12.118473", "stderr": "sed: 1: \"/tmp/academicskills/sit ...\": command a expects \\ followed by text", "stderr_lines": ["sed: 1: \"/tmp/academicskills/sit ...\": command a expects \\ followed by text"], "stdout": "", "stdout_lines": []}
failed: [jbickar] (item=sed -i 's/^.*acsf_site_id/acsf_site_id/g' /tmp/jbickar/sites_json_raw) => {"changed": true, "cmd": "sed -i 's/^.*acsf_site_id/acsf_site_id/g' /tmp/jbickar/sites_json_raw", "delta": "0:00:00.008270", "end": "2018-10-02 17:13:12.151591", "item": "sed -i 's/^.*acsf_site_id/acsf_site_id/g' /tmp/jbickar/sites_json_raw", "msg": "non-zero return code", "rc": 1, "start": "2018-10-02 17:13:12.143321", "stderr": "sed: 1: \"/tmp/jbickar/sites_json ...\": invalid command code j", "stderr_lines": ["sed: 1: \"/tmp/jbickar/sites_json ...\": invalid command code j"], "stdout": "", "stdout_lines": []}
failed: [academicskills] (item=sed -i  "s/\,\"acsf_db_name.*$//g" /tmp/academicskills/sites_json_raw) => {"changed": true, "cmd": "sed -i  \"s/\\,\\\"acsf_db_name.*$//g\" /tmp/academicskills/sites_json_raw", "delta": "0:00:00.008341", "end": "2018-10-02 17:13:12.403873", "item": "sed -i  \"s/\\,\\\"acsf_db_name.*$//g\" /tmp/academicskills/sites_json_raw", "msg": "non-zero return code", "rc": 1, "start": "2018-10-02 17:13:12.395532", "stderr": "sed: 1: \"/tmp/academicskills/sit ...\": command a expects \\ followed by text", "stderr_lines": ["sed: 1: \"/tmp/academicskills/sit ...\": command a expects \\ followed by text"], "stdout": "", "stdout_lines": []}
failed: [jbickar] (item=sed -i  "s/\,\"acsf_db_name.*$//g" /tmp/jbickar/sites_json_raw) => {"changed": true, "cmd": "sed -i  \"s/\\,\\\"acsf_db_name.*$//g\" /tmp/jbickar/sites_json_raw", "delta": "0:00:00.008719", "end": "2018-10-02 17:13:12.429741", "item": "sed -i  \"s/\\,\\\"acsf_db_name.*$//g\" /tmp/jbickar/sites_json_raw", "msg": "non-zero return code", "rc": 1, "start": "2018-10-02 17:13:12.421022", "stderr": "sed: 1: \"/tmp/jbickar/sites_json ...\": invalid command code j", "stderr_lines": ["sed: 1: \"/tmp/jbickar/sites_json ...\": invalid command code j"], "stdout": "", "stdout_lines": []}
failed: [academicskills] (item=sed -i  "s/^acsf_site_id\"\://g" /tmp/academicskills/sites_json_raw) => {"changed": true, "cmd": "sed -i  \"s/^acsf_site_id\\\"\\://g\" /tmp/academicskills/sites_json_raw", "delta": "0:00:00.008446", "end": "2018-10-02 17:13:12.672609", "item": "sed -i  \"s/^acsf_site_id\\\"\\://g\" /tmp/academicskills/sites_json_raw", "msg": "non-zero return code", "rc": 1, "start": "2018-10-02 17:13:12.664163", "stderr": "sed: 1: \"/tmp/academicskills/sit ...\": command a expects \\ followed by text", "stderr_lines": ["sed: 1: \"/tmp/academicskills/sit ...\": command a expects \\ followed by text"], "stdout": "", "stdout_lines": []}
failed: [jbickar] (item=sed -i  "s/^acsf_site_id\"\://g" /tmp/jbickar/sites_json_raw) => {"changed": true, "cmd": "sed -i  \"s/^acsf_site_id\\\"\\://g\" /tmp/jbickar/sites_json_raw", "delta": "0:00:00.008350", "end": "2018-10-02 17:13:12.697968", "item": "sed -i  \"s/^acsf_site_id\\\"\\://g\" /tmp/jbickar/sites_json_raw", "msg": "non-zero return code", "rc": 1, "start": "2018-10-02 17:13:12.689618", "stderr": "sed: 1: \"/tmp/jbickar/sites_json ...\": invalid command code j", "stderr_lines": ["sed: 1: \"/tmp/jbickar/sites_json ...\": invalid command code j"], "stdout": "", "stdout_lines": []}
```

Not sure what that means yet but starting to debug.

**Comment by jbickar** [comment]
Are you running it on OS X? sed works slightly differently on Linux.

**Comment by sherakama** [comment]
I pushed an update for an OSX friendlier option. Please let me know if it still works for you. According to SO it should.

File: roles/sites-json/tasks/main.yml (hunk 2 above)
Comment (jbickar) [reply roles/sites-json/tasks/main.yml]: Well, nuts.

**Comment by sherakama** [comment]
@jbickar can you confirm new sed syntax works on vbox?

**Comment by jbickar** [comment]
I can, but I'm most concerned that it works on `sites-uat`, which I can't test right now without a bit of makework that I don't feel like doing :/

**Comment by jbickar** [comment]
Fails on `my-stanford-vbox`:

```
failed: [a1das] (item=sed -i ".original" -e "s/^.*acsf_site_id/acsf_site_id/g" /tmp/a1das/sites_json_raw) => {"changed": true, "cmd": "sed -i \".original\" -e \"s/^.*acsf_site_id/acsf_site_id/g\" /tmp/a1das/sites_json_raw", "delta": "0:00:00.002594", "end": "2018-10-02 21:16:24.223681", "item": "sed -i \".original\" -e \"s/^.*acsf_site_id/acsf_site_id/g\" /tmp/a1das/sites_json_raw", "msg": "non-zero return code", "rc": 2, "start": "2018-10-02 21:16:24.221087", "stderr": "sed: can't read .original: No such file or directory", "stderr_lines": ["sed: can't read .original: No such file or directory"], "stdout": "", "stdout_lines": []}
failed: [a1das] (item=sed -i ".original" -e "s/\,\"acsf_db_name.*$//g" /tmp/a1das/sites_json_raw) => {"changed": true, "cmd": "sed -i \".original\" -e \"s/\\,\\\"acsf_db_name.*$//g\" /tmp/a1das/sites_json_raw", "delta": "0:00:00.003446", "end": "2018-10-02 21:16:24.337810", "item": "sed -i \".original\" -e \"s/\\,\\\"acsf_db_name.*$//g\" /tmp/a1das/sites_json_raw", "msg": "non-zero return code", "rc": 2, "start": "2018-10-02 21:16:24.334364", "stderr": "sed: can't read .original: No such file or directory", "stderr_lines": ["sed: can't read .original: No such file or directory"], "stdout": "", "stdout_lines": []}
failed: [a1das] (item=sed -i ".original" -e "s/^acsf_site_id\"\://g" /tmp/a1das/sites_json_raw) => {"changed": true, "cmd": "sed -i \".original\" -e \"s/^acsf_site_id\\\"\\://g\" /tmp/a1das/sites_json_raw", "delta": "0:00:00.003552", "end": "2018-10-02 21:16:24.450299", "item": "sed -i \".original\" -e \"s/^acsf_site_id\\\"\\://g\" /tmp/a1das/sites_json_raw", "msg": "non-zero return code", "rc": 2, "start": "2018-10-02 21:16:24.446747", "stderr": "sed: can't read .original: No such file or directory", "stderr_lines": ["sed: can't read .original: No such file or directory"], "stdout": "", "stdout_lines": []}
```

**Comment by jbickar** [comment]
stackoverflow

**Comment by sherakama** [comment]
Tested on UAT and OSX now

```
sheamck@sites-uat:~$ sed -i.original -e 's/dog/dog/g' /tmp/sedtest
sheamck@sites-uat:~$ cat /tmp/sedtest
cat
bat
dog
hat
mat
man
ned

sheamck@sites-uat:~$ sed -i.original -e 's/^dog//g' /tmp/sedtest
sheamck@sites-uat:~$ cat /tmp/sedtest
cat
bat

hat
mat
man
ned
```

**Comment by jbickar** [comment]
**STOP!**

YOU MUST NOT HOP ON POP!

**Comment by jbickar** [comment]
On `my-stanford-vbox`:

```
ERROR! Syntax Error while loading YAML.
  did not find expected alphabetic or numeric character

The error appears to have been in '/home/vagrant/shared/ansible-playbooks/roles/sites-json/tasks/main.yml': line 60, column 99, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

    - 'sed -i.original -e "s/\,\"acsf_db_name.*$//g" /tmp/{{ inventory_hostname }}/sites_json_raw'
    - 'sed -i.original -e "s/^acsf_site_id\"\://g" /tmp/{{ inventory_hostname }}/sites_json_raw' && rm -- "/tmp/{{ inventory_hostname }}/sites_json_raw.original"
                                                                                                  ^ here
```

**Comment by sherakama** [comment]
Stupid quotes

Try again.

**Comment by jbickar** [comment]
This appears to be working on sites-uat.

**Comment by jbickar** [comment]
And on `my-stanford-vbox`.

Let's do eet.

**Pull request merged**